nixos/modules/services/networking/xray.nix

```
@@ -0,0 +1,95 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+{
+  options = {
+
+    services.xray = {
+      enable = mkOption {
+        type = types.bool;
+        default = false;
+        description = lib.mdDoc ''
+          Whether to run xray server.
+
+          Either `configFile` or `config` must be specified.
+        '';
+      };
+
+      package = mkOption {
+        type = types.package;
+        default = pkgs.xray;
+        defaultText = literalExpression "pkgs.xray";
+        description = lib.mdDoc ''
+          Which xray package to use.
+        '';
+      };
+
+      configFile = mkOption {
+        type = types.nullOr types.path;
+        default = null;
```

```suggestion
        type = types.path;
        default = (pkgs.formats.json {}).generate "config.json" cfg.settings;
        defaultText = "A configuration file generated from `services.xray.settings`."
```

Optionally you can add back in validation, but I wanted a simple example to show you so I didn't.